### ISSUE TITLE
Spline widget appears behind image mapper

### ISSUE THREAD

**Issue by agirault** [Kitware/vtk-js#1420]
I mixed the `SplineWidget` example with the `ImageMapper` example:

<details><summary>See code</summary>

```js
import 'vtk.js/Sources/favicon';

import vtkFullScreenRenderWindow from 'vtk.js/Sources/Rendering/Misc/FullScreenRenderWindow';
import vtkSplineWidget from 'vtk.js/Sources/Widgets/Widgets3D/SplineWidget';
import vtkWidgetManager from 'vtk.js/Sources/Widgets/Core/WidgetManager';

import { SlicingMode } from 'vtk.js/Sources/Rendering/Core/ImageMapper/Constants';
import vtkRTAnalyticSource from 'vtk.js/Sources/Filters/Sources/RTAnalyticSource';
import vtkImageMapper from 'vtk.js/Sources/Rendering/Core/ImageMapper';
import vtkImageSlice from 'vtk.js/Sources/Rendering/Core/ImageSlice';
import vtkInteractorStyleImage from 'vtk.js/Sources/Interaction/Style/InteractorStyleImage';
import vtkColorTransferFunction from 'vtk.js/Sources/Rendering/Core/ColorTransferFunction';
import vtkPiecewiseFunction from 'vtk.js/Sources/Common/DataModel/PiecewiseFunction';

import controlPanel from './controlPanel.html';

// ----------------------------------------------------------------------------
// Standard rendering code setup
// ----------------------------------------------------------------------------

const fullScreenRenderer = vtkFullScreenRenderWindow.newInstance({
  background: [0, 0, 0],
});
const renderer = fullScreenRenderer.getRenderer();
const renderWindow = fullScreenRenderer.getRenderWindow();

// ----------------------------------------------------------------------------
// ImageMapper example
// ----------------------------------------------------------------------------

const rtSource = vtkRTAnalyticSource.newInstance();
rtSource.setWholeExtent(0, 200, 0, 200, 0, 200);
rtSource.setCenter(100, 100, 100);
rtSource.setStandardDeviation(0.3);

const mapper = vtkImageMapper.newInstance();
mapper.setInputConnection(rtSource.getOutputPort());
mapper.setSliceAtFocalPoint(true);
mapper.setSlicingMode(SlicingMode.Z);
// mapper.setZSlice(5);

const rgb = vtkColorTransferFunction.newInstance();
global.rgb = rgb;
rgb.addRGBPoint(0, 0, 0, 0);
rgb.addRGBPoint(255, 1, 1, 1);

const ofun = vtkPiecewiseFunction.newInstance();
global.ofun = ofun;
ofun.addPoint(0, 1);
ofun.addPoint(150, 1);
ofun.addPoint(180, 0);
ofun.addPoint(255, 0);

const actor = vtkImageSlice.newInstance();
actor.getProperty().setColorWindow(255);
actor.getProperty().setColorLevel(127);
// Uncomment this if you want to use a fixed colorwindow/level
// actor.getProperty().setRGBTransferFunction(rgb);
actor.getProperty().setPiecewiseFunction(ofun);
actor.setMapper(mapper);
renderer.addActor(actor);

const iStyle = vtkInteractorStyleImage.newInstance();
iStyle.setInteractionMode('IMAGE_SLICING');
renderWindow.getInteractor().setInteractorStyle(iStyle);

const camera = renderer.getActiveCamera();
const position = camera.getFocalPoint();
// offset along the slicing axis
const normal = mapper.getSlicingModeNormal();
position[0] += normal[0];
position[1] += normal[1];
position[2] += normal[2];
camera.setPosition(...position);
switch (mapper.getSlicingMode()) {
  case SlicingMode.X:
    camera.setViewUp([0, 1, 0]);
    break;
  case SlicingMode.Y:
    camera.setViewUp([1, 0, 0]);
    break;
  case SlicingMode.Z:
    camera.setViewUp([0, 1, 0]);
    break;
  default:
}
camera.setParallelProjection(true);
renderer.resetCamera();
renderWindow.render();

// ----------------------------------------------------------------------------
// SplineWidget example
// ----------------------------------------------------------------------------

const widgetManager = vtkWidgetManager.newInstance();
widgetManager.setRenderer(renderer);

const widget = vtkSplineWidget.newInstance();

widgetManager.addWidget(widget);

renderer.resetCamera();
widgetManager.enablePicking();

// -----------------------------------------------------------
// UI control handling
// -----------------------------------------------------------

fullScreenRenderer.addController(controlPanel);

document.querySelector('button').addEventListener('click', () => {
  widgetManager.grabFocus(widget);
});

```
</details>

The spline handles don't stick on the actor surface here (the spline doesn't look curvy because I had to press `Shift` to take a screenshot, which modifies the spline representation):
<img width="611" alt="Screen Shot 2020-04-27 at 7 12 40 PM" src="https://user-images.githubusercontent.com/4910855/80430047-7e73c680-88bb-11ea-8571-7155874958c0.png">

cc: @RemiCecchinato, @finetjul 

Thank you

Edit: second issue moved to #1423

**Comment by finetjul** [comment]
I believe the first issue is the same as in the [ShapeWidget](https://kitware.github.io/vtk-js/examples/ShapeWidget.html) example.

I do not know when this has been broken (@boucaud could it be your recent changes ?)

This can be solved (in ShapeWidget example at least) with:
```
image.imageMapper.setResolveCoincidentTopologyToPolygonOffset();
image.imageMapper.setRelativeCoincidentTopologyPolygonOffsetParameters(
      1.0,
      1.0
    );
```

However, I do not think it is the right solution. I think Widgets should have coincidentTopolyogy parameters set. What do you think @floryst @boucaud ?

**Comment by boucaud** [comment]
After testing, I can say the first issue is due to my changes to the ImageMapper. I'm not sure about the crash though.

**Comment by boucaud** [comment]
> However, I do not think it is the right solution. I think Widgets should have coincidentTopolyogy parameters set. What do you think @floryst @boucaud ?

It makes sense (to me) to have all widget mappers' coincidentTopolyogy parameters set to some negative value, as they should always be above the other mappers.

**Comment by finetjul** [comment]
> After testing, I can say the ShapeWidget issue is due to my changes to the ImageMapper. 

Is it a bug then ? Is there a more general fix to do ?

**Comment by boucaud** [comment]
> Is it a bug then ? Is there a more general fix to do ?

There is a bug with the SplineWidget not setting coincident topology parameters. It also happens with other widgets. This is easily solved by setting those parameters.

However there is another problem in this example. `mapper.setSliceAtFocalPoint(true);` causes the imageMapper's quad to be rendered with negative Z coordinates, so it will always overlap the widget which is rendered at z=0

**Comment by boucaud** [comment]
> Is it a bug then ? Is there a more general fix to do ?

There is a bug with the SplineWidget not setting coincident topology parameters. It also happens with other widgets. This is easily solved by setting those parameters.

However there is another problem in this example. `mapper.setSliceAtFocalPoint(true);` causes the imageMapper's quad to be rendered with negative Z coordinates, so it will always overlap the widget which is rendered at z=0

**Comment by finetjul** [comment]
> After testing, I can say the ShapeWidget issue is due to my changes to the ImageMapper. 

But your change was not supposed to change anything by default, was it ? And because Widgets do not set coincident topolygy parameters nor does the imageMapper, why is there a regression ?

**Comment by boucaud** [comment]
> But your change was not supposed to change anything by default, was it ? And because Widgets do not set coincident topolygy parameters nor does the imageMapper, why is there a regression ?

The shape widgets do enable coincident topology (through vtkCircleContextRepresentation).

The issue with the SplineWidget being behind the image mapper is also present before my change.

**Comment by finetjul** [comment]
> The shape widgets do enable coincident topology (through vtkCircleContextRepresentation).

Gotcha, that makes sense then. Your recent change has actually fixed inconsistent behavior. It revealed some bugs.
So there are 2 issues to fix:
  - find a way to automatically set coincident topology on all widget actors. @floryst  any suggestion ?
  - for the SplineWidget Z issue, I agree, it should not be handled via coincident topologoy parameters but the Z coordinates must be modified (by the application ? the widget?). I think here there are multiple behaviors possibles (if scrolling happens before, while or after placing spline points).

**Comment by boucaud** [comment]
One solution that would fix those issues (and remove the need for coincident topology) would be to render all widgets in a separate framebuffer, and superpose that to the renderbuffer, but that may not be realistic.

**Comment by agirault** [comment]
> However there is another problem in this example. mapper.setSliceAtFocalPoint(true); causes the imageMapper's quad to be rendered with non-zero coordinates, so it will overlap the widget which is rendered at z=0.

FYI, in my use case, I won't be using `setSliceAtFocalPoint`, but do data slicing (I, J or K) and align the camera with it. Something like that:

<details><summary>See code snippet</summary>

```js
// get image direction to realign camera
const dir9 = imageData.getDirection();
const dir3x3 = [
  [dir9[0], dir9[3], dir9[6]],
  [dir9[1], dir9[4], dir9[7]],
  [dir9[2], dir9[5], dir9[8]],
];

// compute the focal point
renderer.resetCamera();

// offset along the slicing axis (-K)
const position = camera.getFocalPoint();
const normal = [0, 0, -1];
multiply3x3_vect3(dir3x3, normal, normal);
position[0] += normal[0];
position[1] += normal[1];
position[2] += normal[2];
camera.setPosition(...position);

// set view up (-J)
const up = [0, -1, 0];
multiply3x3_vect3(dir3x3, up, up);
camera.setViewUp(up);

// reset bounds
renderer.resetCamera();
renderer.resetCameraClippingRange();

// parallel projection
camera.setParallelProjection(true);

// change slice
mapper.setSlice(s);
renderWindow.render();
```
</details>

> the Z coordinates must be modified (by the application ? the widget?). I think here there are multiple behaviors possibles (if scrolling happens before, while or after placing spline points).

FYI, in my use case, I'd like to use the spline as a representation of a segmented 3d image on top of the original 3d image, with the spline components defined for each slice at a different Z coordinate. From what I can tell, I could either have:
- a widget for each slice that coincides with the slices + an offset, or...
- a single widget for all slices that is between the camera and the whole dataset, with different widget states being loaded every time the slice changes (I was leaning toward that one).

Not sure if anything I said here is helpful to you, but I thought I'd clarify what use case I'm looking for.

Thanks

**Comment by finetjul** [comment]
@floryst what do you think? should Coincident Topology be generalized to Widgets or is there another mecanism ?

**Comment by floryst** [comment]
Right now there isn't any other standard mechanism for widgets to show up in front. Having coincident topology for widget representations in general, esp the 2D ones, would be very useful.

**Comment by agirault** [comment]
The second part of the issue was moved to #1423

**Comment by agirault** [comment]
Fixed by #1430, thank you @boucaud @finetjul
